### ISSUE TITLE
Decide if we need to speed up adding subtrees for existing database blocks

### ISSUE THREAD

**Issue by arya2** [ZcashFoundation/zebra#7441]
# TODO

- [ ] Make sure the current code is correct
- [ ] Work out if any performance upgrades are needed
- [ ] Split this ticket into multiple tickets, based on which performance upgrades we decide to do.

## Motivation

The database format upgrade for adding subtrees is slow on mainnet. It takes about 20 minutes on a recent machine.

This is already an acceptable time for wallet users to wait for a once-off upgrade, so this ticket is low or optional priority.

### Design Alternatives

Together, these design alternatives could reduce the upgrade time to a few minutes, which is how long it takes Zebra to start up anyway.

These alternatives are sorted in order of complexity. If we do each optimisation in a separate PR, and check the time after each one, we'll know if it worked, and when we can stop doing upgrades.

Many of these optimisations are from the `zcashd` PR in our ticket #6953: https://github.com/zcash/zcash/pull/6694/files#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2R5815

#### Start at the relevant Network Upgrade

It isn't possible to have 

#### Only Read Transactions Ticket

If this fix turns out to be complicated, we should do profiling (for example, a flamegraph) to confirm if it's needed, before we write any complex code.

If reading blocks from the database is slow, we could skip reading the block headers, and only read transactions. There is already a database method (or database code in the existing block method) for this.

We should time the upgrade after this change, and if it isn't any faster, we shouldn't merge the change. It's possible that it's reading the transactions or calculating the note commitments that's the slow part.

#### Incremental Upgrade Ticket

If we start the upgrade from the first missing subtree, then we don't waste the work done by previously interrupted upgrades.

This reduces the time by the time spent on previous incomplete upgrades.

Implementing this ticket makes testing other changes more complicated, so we might want to do it last. (If we do incremental upgrades, they skip updating some subtrees, so our validity tests won't be testing the latest code any more.)

#### Testing Even Indexes Ticket & Re-Using Even Ommers Ticket

This depends on ticket #7446.

Before we start this change, we also need to add another 8 tests for fixed subtrees from `zcashd`:
* Odd subtree indexes & even subtree indexes

Because of the way the frontier is structured, the ommers of incomplete odd-numbered subtrees contain the root of the previous even-numbered subtree. The subtree's root stays in the ommers until the odd-numbered subtree is completed.

So if we get the even-numbered subtree roots from the ommers, we don't need to do any calculation, or append any notes.

This reduces the time by approximately half.

## Future Work / Other Tickets

These upgrades are more complicated and might need to be scheduled in separate tickets.

#### Parallel Subtree Calculation PR

<details>

<summary>optional upgrade - technically complex</summary>

Spawn rayon tasks to get roots of subtrees ending in blocks that started new subtrees.

This divides the time by approximately the number of CPUs.

This upgrade is technically complex, because we need to handle cancellation and panics across the threads correctly, by:
- using rayon APIs that resume unwinds in the calling thread
- switching to a `cancel_receiver` type that can be cloned (`tokio::sync::watch` and our own `WatchReceiver` wrapper)

</details>

**Comment by arya2** [comment]
> Because of the way the frontier is structured, the even-numbered subtree roots are available as one of the frontier ommers until the next subtree is created.

I don't think calculating the root is slow by comparison to appending notes, it seems to call `H::combine` ~2x as many times as `append()`, but we're only getting <2k roots for the current state.

**Comment by teor2345** [comment]
> > Because of the way the frontier is structured, the even-numbered subtree roots are available as one of the frontier ommers until the next subtree is created.
> 
> I don't think calculating the root is slow by comparison to appending notes, it seems to call `H::combine` ~2x as many times as `append()`, but we're only getting <2k roots for the current state.

As far as I understand, the ommers of incomplete odd-numbered subtrees contain the root of the previous even-numbered subtree. So if we want to get these subtree roots during the upgrade, we don't need to do any calculation, or append any notes.

**Comment by teor2345** [comment]
I'm running the same kind of upgrade on `zcashd` right now. It looks like it will take at least 10 hours, and none of the subtrees will be available until it has finished.

So Zebra is already significantly faster than `zcashd`, and #7407 will make Zebra appear even faster to users. (Our current performance goal is "not significantly slower than `zcashd`".)

**Comment by teor2345** [comment]
> I'm running the same kind of upgrade on `zcashd` right now. It looks like it will take at least 10 hours, and none of the subtrees will be available until it has finished.

`zcashd` takes 7 hours to reindex and about ~18 hours to rebuild the chain state database.

**Comment by teor2345** [comment]
There's no point in speeding up the upgrade, because it was in the last release, so it is finished for most users.
